### PR TITLE
Fix issues in BTS-1255

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix most issues described in https://arangodb.atlassian.net/browse/BTS-1255

* BTS-1255: Fix sporadic memory usage accounting underflows in in-memory cache subsystem. Also turn the algorithm for freeing memory from a cache's buckets from a non-deterministic one that did not guarantee progress into a bounded algorithm with guaranteed progress.

  Also enable 2 previously disabled tests because they are now stable after the fixes.

  Additionally fix a bug in the cache that prevented cached items to be found after data in a bucket got evicted due to memory pressure. Previously some items could not be found after eviction although they were still present in the cache.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [x] Backport for 3.11: https://github.com/arangodb/arangodb/pull/19094
  - [x] Backport for 3.10: https://github.com/arangodb/arangodb/pull/19095
  - [ ] Backport for 3.9: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/BTS-1255
- [ ] Design document:

### CODE REVIEW

**Review by kvahed — APPROVED** [review]
LGTM

**Review by mpoeter — CHANGES_REQUESTED** [review]
I really don't like the code duplication the whole byId/byHash separation introduces. Can we avoid that?

File: arangod/Cache/Manager.cpp [hunk 1]
```diff
@@ -723,16 +732,13 @@ void Manager::resizeCache(Manager::TaskEnvironment environment,
...
-      TRI_ASSERT(_globalAllocation >= _fixedAllocation);
-      _peakGlobalAllocation =
-          std::max(_globalAllocation, _peakGlobalAllocation);
-    }
+    TRI_ASSERT(_globalAllocation + newLimit - oldLimit >= _fixedAllocation);
+    _globalAllocation += newLimit;
+    _globalAllocation -= oldLimit;
+    TRI_ASSERT(_globalAllocation >= _fixedAllocation);
+    _peakGlobalAllocation = std::max(_globalAllocation, _peakGlobalAllocation);
+    metaGuard.release();
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/Manager.cpp]: Fixed

File: arangod/Cache/Manager.cpp [hunk 2]
```diff
@@ -213,7 +213,15 @@ void Manager::shutdown() {
       SpinUnlocker unguard(SpinUnlocker::Mode::Write, _lock);
       cache->shutdown();
     }
+
+    TRI_ASSERT(_activeTables == 0);
     freeUnusedTables();
+
+    for (auto& it : _tables) {
+      TRI_ASSERT(it.empty());
+    }
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/Manager.cpp]: Fixed

File: arangod/Cache/Cache.cpp [hunk 3]
```diff
@@ -397,30 +397,29 @@ bool Cache::freeMemory() {
...
-        }
+      bool underLimit = reclaimMemory(reclaimed);
+      if (underLimit) {
+        // we have free enough memory.
+        // don't continue
+        return false;
       }
     }
-  }
+    if (++attempts % 32 == 0 && isShutdown()) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/Cache.cpp]: Fixed

File: arangod/Cache/PlainCache.cpp [hunk 4]
```diff
@@ -220,38 +221,57 @@ PlainCache<Hasher>::~PlainCache() {
...
+    if (!cb(reclaimed)) {
+      break;
     }
   }
 
-  return reclaimed;
+  if (maybeMigrate) {
+    // caution: calling idealSize() can have side effects
+    // and trigger a table growth!
+    requestMigrate(table->idealSizeValue());
```
[the remark this answers is not in the document]
Comment (mpoeter) [reply arangod/Cache/PlainCache.cpp]: That's fine, I was just wondering whether we can find a better name that actually describes what the function is doing (including its side effects).

[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/PlainCache.cpp]: I have been looking at the existing `idealSize()` method for 5 minutes now and cannot come up with a better name.
It is unfortunate that it has side-effects, but I decided against changing that for this PR.
I can't come up with a better name right now, unfortunately, so I vote for leaving the name as is for now.
There are code comments on each call site now that warn about the call's potential side-effects.

**Review by mpoeter — APPROVED** [review]

**Review by kvahed — APPROVED** [review]
LGTM

**Pull request merged**